pkg/services/object/metrics.go

```diff
@@ -0,0 +1,159 @@
+package object
+
+import (
+	"context"
+	"time"
+
+	"github.com/nspcc-dev/neofs-api-go/v2/object"
+	"github.com/nspcc-dev/neofs-node/pkg/services/util"
+)
+
+type (
+	MetricCollector struct {
+		next    ServiceServer
+		metrics MetricRegister
+	}
+
+	getStreamMetric struct {
+		util.ServerStream
+		stream  GetObjectStream
+		metrics MetricRegister
+	}
+
+	putStreamMetric struct {
+		stream  object.PutObjectStreamer
+		metrics MetricRegister
+	}
+
+	MetricRegister interface {
+		IncGetReqCounter()
+		IncPutReqCounter()
+		IncHeadReqCounter()
+		IncSearchReqCounter()
+		IncDeleteReqCounter()
+		IncRangeReqCounter()
+		IncRangeHashReqCounter()
+
+		AddGetReqDuration(time.Duration)
+		AddPutReqDuration(time.Duration)
+		AddHeadReqDuration(time.Duration)
+		AddSearchReqDuration(time.Duration)
+		AddDeleteReqDuration(time.Duration)
+		AddRangeReqDuration(time.Duration)
+		AddRangeHashReqDuration(time.Duration)
+
+		AddPutPayload(int)
+		AddGetPayload(int)
+	}
+)
+
+func NewMetricCollector(next ServiceServer, register MetricRegister) *MetricCollector {
+	return &MetricCollector{
+		next:    next,
+		metrics: register,
+	}
+}
+
+func (m MetricCollector) Get(req *object.GetRequest, stream GetObjectStream) error {
+	t := time.Now()
+	defer func() {
+		m.metrics.IncGetReqCounter()
+		m.metrics.AddGetReqDuration(time.Since(t))
+	}()
+
+	return m.next.Get(req, &getStreamMetric{
+		ServerStream: stream,
+		stream:       stream,
+		metrics:      m.metrics,
+	})
+}
+
+func (m MetricCollector) Put(ctx context.Context) (object.PutObjectStreamer, error) {
+	t := time.Now()
+	defer func() {
+		m.metrics.IncPutReqCounter()
+		m.metrics.AddPutReqDuration(time.Since(t))
```

Comment:
I think we should `AddPutReqDuration` in `putStreamMetric` calls too, current call does not transmit any data.

Reply:
Probably it is better to drop duration metric in `Get` and `Put` and use it in `Send` functions in the streams. What do you think?

Reply:
The only call that doesn't transmit the data is `Put`, so this meaningfully only for it.

Reply:
> The only call that doesn't transmit the data is Put, so this meaningfully only for it.

I see, thank you. I moved `AddPutReqDuration` to `CloseAndRecv` since it is last invocation call.